### PR TITLE
Fix for Update Global Offline Indicator

### PR DESCRIPTION
**Pull request by jasperhuangg**
_NOTE TO REVIEWER: Recommend viewing [diff without whitespace changes](https://github.com/Expensify/App/pull/10140/files?diff=split&w=1) for more succinct review_

We previously had [this PR](https://github.com/Expensify/App/pull/9895) which was merged, but was later reverted due to a number of layout issues associated with the new KeyboardAvoidingView usage in the ScreenWrapper. This PR adds back everything that was reverted, and fixes the behavior of the ScreenWrapper in a number of places on Android. See linked issues in the old PR and the issue below for more context.

There were two main issues that we were running into that were causing bizarre gaps to form between the keyboard and the actual content. At times these gaps would push up the content so high that it would get cut off:

1. On Android, we were using KeyboardAvoidingView components that had `'padding'` passed into the `behavior` prop instead of `'height'`. It's known that KeyboardAvoidingViews can have unexpected behavior if you don't use `'height'`.
https://user-images.githubusercontent.com/93399543/180846058-039d8bd0-d5e5-4d94-8ae1-9cb8e6c818bd.mp4
2. On iOS, some components had KeyboardSpacer nested inside KeyboardAvoidingView, which would reapply the padding for the KeyboardAvoidingView twice, causing the gap to occur.

I've addressed both of these issues in this PR, primarily by refactoring the ScreenWrapper component (which houses the KeyboardAvoidingView).

### Fixed Issues

$ https://github.com/Expensify/Expensify/issues/218715

### Tests

- [x] Verify that no errors appear in the JS console
See QA steps

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The Contributor+ will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/docs/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/docs/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/docs/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/docs/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### QA Steps
1. Sign into an account on NewDot.
2. Turn off your wifi.
3. Verify that the OfflineIndicator appears correctly (see screen captures).
6. Turn on your wifi.
7. Verify that the OfflineIndicator goes away.
8. Repeat on all platforms.

- [ ] Verify that no errors appear in the JS console

#### Web
Screen Shot 2022-08-01 at 7 33 26 PM

#### Mobile Web
ios safari
android chrome

#### Desktop
Screen Shot 2022-08-01 at 7 36 08 PM

#### iOS
Screen Shot 2022-08-01 at 7 28 56 PM

#### Android
Screen Shot 2022-08-01 at 7 30 31 PM

### CODE REVIEW

**Comment by arosiclair** [comment]
Code seems good, but I'm having trouble following all of the issues that came up with our first go at this. Can you summarize those issues and how to test them so we can make sure we've got it right this go around?

**Comment by jasperhuangg** [comment]
@arosiclair got it! Will update the OP

**Comment by jasperhuangg** [comment]
@arosiclair There seems to be a bug on Android that I found on `main` that you'll probably experience in this PR:

https://expensify.slack.com/archives/C03TQ48KC/p1659115536665949

**Comment by neil-marcellini** [comment]
This is a big PR and it's blocking a lot so I thought I would jump in and help review. Please fix those merge conflicts first.

File: src/components/ScreenWrapper.js [hunk 1]
```diff
@@ -1,147 +0,0 @@
...
-    includePaddingTop: PropTypes.bool,
-
-    // Called when navigated Screen's transition is finished.
-    onTransitionEnd: PropTypes.func,
-
-    // react-navigation navigation object available to screen components
-    navigation: PropTypes.shape({
-        // Method to attach listener to Navigation state.
-        addListener: PropTypes.func.isRequired,
-    }),
```
Comment (neil-marcellini) [inline src/components/ScreenWrapper.js:42]: NAB: I don't see these propTypes in BaseScreenWrapper. Is withNavigation is somehow applying the navigation prop type. If so, how does that work?

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
I reviewed all the files and left 2 comments. Now I'm going to test all the pages that were changed, on all platforms. I'll also test out the offline indicator. Please add test and QA steps to the issue.

File: src/components/ScreenWrapper.js (hunk 1 above)
Comment (jasperhuangg) [reply src/components/ScreenWrapper.js:42]: Yes, `withNavigation` automatically applies those propTypes. I didn't include them in the propTypes because we don't include them anywhere else we use `withNavigation`, for instance [here](https://github.com/Expensify/App/blob/631994f7b9a974ca3ce92b8405bd1416a9d34623/src/components/SettlementButton.js#L97) with the SettlementButton.

Comment (neil-marcellini) [reply src/components/ScreenWrapper.js:42]: Yeah I noticed that the prop types aren't used elsewhere and I'm wondering why that is and if / how they are applied at all. I'll ask in Slack for my own curiosity.

**Review by neil-marcellini — COMMENTED** [review]
~~I'm not seeing the offline indicator on Android for some reason.~~ (Edit: turn off mobile data)
image

**Comment by neil-marcellini** [comment]
Hmm, it works fine on iOS.

**Comment by neil-marcellini** [comment]
I was able to get the android offline indicator to show by going to Settings > Network and internet. Turn off wifi. Then go to Mobile network and turn off Mobile data.

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
cc @Expensify/design
Compared to the mockups [here](https://docs.google.com/document/d/1Ija9CHiNBH1HVnnITA-XuohF31BoiseY18a8_BxFtjc/edit#heading=h.ndwpocqbnb99) the offline indicator has too much padding above it on mobile. I know there was some discussion about that on your last PR, but I think this is fixable and necessary to make it look good.

image

On desktop the indicator needs to be pushed to the left. It should have the same distance between it and the left edge as on mobile.
image

**Comment by michelle-thompson** [comment]
Yup, I agree with the suggestions that @neil-marcellini made above

**Comment by jasperhuangg** [comment]
@neil-marcellini Thanks so much for taking the time test out all those views! Fixed the stuff you were talking about, let me know if there's anything else:
Screen Shot 2022-08-01 at 7 36 08 PM
Screen Shot 2022-08-01 at 7 33 26 PM
Screen Shot 2022-08-01 at 7 30 31 PM
Screen Shot 2022-08-01 at 7 28 56 PM

File: src/pages/home/ReportScreen.js [hunk 2]
```diff
@@ -181,6 +182,15 @@ class ReportScreen extends React.Component {
...
 
+    /**
+     * @param {SyntheticEvent} e
+     */
+    updateViewportOffsetTop(e) {
+        const viewportOffsetTop = lodashGet(e, 'target.offsetTop', 0);
+        this.setState({viewportOffsetTop});
+    }
+
+
```
Comment (neil-marcellini) [inline src/pages/home/ReportScreen.js:193]: Extra blank line here causing a linting error.

**Review by neil-marcellini — CHANGES_REQUESTED** [review]

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Here's what I get when offline on iOS, the spacing isn't fixed. It looks like the min height is getting set correctly, so something else must be causing it.

Aha, so if we show the recipient local time then we also apply this style that has `minHeight: 90`. That condition is probably why you didn't see the problem.
https://github.com/Expensify/App/blob/4df147c720255afc2b73a752793965be89fab55e/src/pages/home/report/ReportActionCompose.js#L499

https://github.com/Expensify/App/blob/2518e21a658c3ca30868350dbb16b6c0de33bff7/src/styles/styles.js#L1341-L1343



With a red border on `chatItemComposeWithFirstRow:

**Comment by neil-marcellini** [comment]
Sorry to bring up another headache, but the full screen composer no longer works properly on iOS after these changes.

On your branch:

https://user-images.githubusercontent.com/26260477/182424678-3958f47b-a004-4e27-8bda-79a06b6cf066.mp4


On main:

https://user-images.githubusercontent.com/26260477/182424703-8af456a7-8784-49a1-9881-63dcb2fa8d20.mp4

**Comment by shawnborton** [comment]
Also just wanted to confirm that we don't get any screen bouncing when the offline indicator shows up and disappears? Ideally this takes up the exact same space/spot as the "User is typing" indicator.

**Comment by jasperhuangg** [comment]
> Also just wanted to confirm that we don't get any screen bouncing when the offline indicator shows up and disappears? Ideally this takes up the exact same space/spot as the "User is typing" indicator.

Ah there is a very slight amount of jumping, I'll make sure to get rid of it.

**Comment by jasperhuangg** [comment]
@neil-marcellini fixed the issues you mentioned!

Screen Shot 2022-08-02 at 4 01 03 PM
Screen Shot 2022-08-02 at 4 49 43 PM
Screen Shot 2022-08-02 at 4 45 53 PM


also @shawnborton made sure that there isn't any jumping happening:

https://user-images.githubusercontent.com/31285285/182495850-6b30dde9-3dd7-471a-bd54-d0fbd3485682.mp4

**Comment by shawnborton** [comment]
Awesome, thank you!

File: src/CONST.js [hunk 3]
```diff
@@ -477,6 +477,7 @@ const CONST = {
     EMOJI_PICKER_ITEM_HEIGHT: 40,
     EMOJI_PICKER_HEADER_HEIGHT: 38,
     COMPOSER_MAX_HEIGHT: 125,
+    CHAT_FOOTER_MIN_HEIGHT: 65,
```
Comment (shawnborton) [inline src/CONST.js:480]: Why 65 out of curiosity? We're big fans of numbers divisible by 4 so just wondering :)

Comment (jasperhuangg) [reply src/CONST.js:480]: this was actually taken from [here](https://github.com/Expensify/App/pull/10140/files#diff-8ab8ba45ab927f1430b8d1209d187e003a9cdfe476f809e4483ef64cdba253ebL1370)

**Review by ctkochan22 — APPROVED** [review]

File: src/CONST.js (hunk 3 above)
Comment (shawnborton) [reply src/CONST.js:480]: Ah cool if it already exists, no worries then. Thanks for listening!

**Review by shawnborton — APPROVED** [review]

**Review by mountiny — APPROVED** [review]
Tested this nad looks good to me, the only problem I had was testing on iOS where it took a long while before the offline indicator showed up and then it kinda randomly dissappeared twice, probably not sure if the device is offline or not.

Not sure what is the best way to set the ios simulator offline. Styling wise it looks good to me!

**Review by neil-marcellini — APPROVED** [review]
Thanks for the changes, it's working really well now! I think it's good to go.

NAB:
It would be nice to move the mobile `ReportTypingIndicator` to the right side so that it doesn't stack on top of the offline indicator. It would also match the web / desktop behavior. Doing that might be kind of tricky, and I doubt many people hit the max comment length, so I think we can do it later.

Mobile:


Web:
image

**Comment by jasperhuangg** [comment]
> It would be nice to move the mobile `ReportTypingIndicator` to the right side so that it doesn't stack on top of the offline indicator.

We can always make an issue for later? @neil-marcellini

**Comment by mananjadhav** [comment]
I am tagging this PR to highlight an issue fixed [here](https://github.com/Expensify/App/issues/16683). All conditions in ternary expressions or left-hand operands on conditional renders, should be boolean. This PR is one of the PRs that uses conditional render with string operands, hence I am tagging it here for the contributors to check. 

We've also updated the item in the checklist with this [PR](https://github.com/Expensify/App/pull/18069) to avoid this issue in the future.
